design-proposals/persist-vmi-firmware-uuid.md

```diff
@@ -0,0 +1,185 @@
+
+**Pros:**
+- Straightforward implementation.
+- Avoids the need to introduce new API fields.
+- Fully backward compatible.
+
+**Cons:**
+- Potential misuse of the spec field, which ideally should only reflect the user’s intended configuration.
+- The majority of users may not consider the UUID part of the desired VM state, making this field irrelevant from their perspective.
+- Can lead to longer and more complex VM definitions, impacting readability and management.
+
+---
+
+### 2. Generate a Default UUID via Webhook and Set it to the Spec Field
+
+**Description:** A webhook generates a UUID if the user does not provide one. This UUID is then set in the VM's spec field.
+
+**Pros:**
+- Retains backward compatibility and avoids introducing new API fields.
+
+**Cons:**
+- Same concerns as above about the spec field.
+- Using webhooks to assign defaults can degrade performance, especially with high volumes of VMs, as `virt-api` may become a bottleneck.
+- May compromise scalability, as generating a UUID via webhook adds overhead when multiple VMs start simultaneously.
+
+---
+
+### 3. Create a New Firmware UUID Field in VM Status
+
+**Description:** A new field is added under VM status to store the generated firmware UUID upon first VM boot.
+
+**Pros:**
+- Keeps the spec clean, avoiding unnecessary fields in the user-defined configuration.
+- Provides a clear separation between user configuration (spec) and system-generated information (status).
+
+**Cons:**
+- Introduces a new API field, which could increase API surface area and requires long-term support.
+- New fields may impact system performance and resource load.
+- Adds to API complexity and may reduce readability.
+
+*Note:* To minimize this con, consider setting the UUID as a condition under VM status instead of introducing a separate field.
+
+---
+
+### 4. Introduce a Breaking Change to Ensure Universal Uniqueness
+
+**Description:** Modify UUID generation to a combination of `VM.Name`, `vm.Namespace`, and a random hash
+to ensure that UUIDs are universally unique, even across clusters.
+
+Users would be notified of this change in advance through logs or alerts,
+and optional tooling could be provided to help users preserve their current firmware UUIDs by adding them to the spec.
+
+**Pros:**
+- Ensures universally unique UUIDs, meeting the UUID standard and avoiding conflicts
+  when migrating VMs across clusters.
+- Keeps the spec clean without repurposing fields.
+- Simple to implement without introducing new API fields.
+
+**Cons:**
+- This approach is a breaking change, requiring user intervention to preserve existing UUIDs.
+- Requires additional communication and support, as users may need to update
+  or reconfigure their VMs to prevent disruptions.
+- May necessitate tooling to facilitate UUID preservation for compatibility with existing workloads.
+
+
+### 5. Upgrade-Specific Persistence of Firmware UUID
```

Comment:
I don’t necessarily object to this approach—in fact, it’s likely unavoidable to prevent disruption to existing workloads—but the process for implementing it should be more thoroughly detailed IMO.

Correct me if I'm wrong, but if VMs are stopped during the upgrade, they cannot rely on VMI objects to derive and persist the firmware UUID, as those objects won’t exist. As a result, existing VMs without a defined `spec.firmware.uuid` field would need to populate the persistent UUID field (whatever that may be) using the current method (based on the name). The new method should then be applied exclusively to new VMs.

While this doesn’t cover _all_ use cases, it’s reasonable to assume it addresses the majority—except for scenarios involving (for example) backup and restore, which present a separate set of challenges.

Reply:
I don't understand the reservation that you are raising. Option 5 does not distinguish between running and non-running VMs. During upgrade, they are all changed to persist the buggy uuid. After upgrade, all VMs receive a fresh uuid.

The problem of restoring old VMs that did not persist their buggy uuid sounds important, but not specific to this option. @dasionov can you address it? in the design? I don't have any good idea...

Reply:
IIRC the original idea was to persist the UUID in existing VMs by obtaining it from the VMIs, therefore my comment. Perhaps I mixed up the proposal with implementation, my bad, you can disregard the comment in that case.